### ISSUE TITLE
expr - error: couldn't create 

### ISSUE THREAD

**Issue by 0xJ1** [libpd/libpd#112]
I'm getting an issue where expr cannot be created. log:

D/MainActivity.pdPost: expr
D/MainActivity.pdPost:  
D/MainActivity.pdPost: $f1
D/MainActivity.pdPost:  
D/MainActivity.pdPost: *
D/MainActivity.pdPost:  
D/MainActivity.pdPost: (100.0/127.0)
D/MainActivity.pdPost: error: ... couldn't create

[libpd_controls_test.zip](https://github.com/libpd/libpd/files/103418/libpd_controls_test.zip)

**Comment by danomatika** [comment]
Looks like you're using the C library directly as the prints aren't compiled on 1 line (which is the default). 

Check the Makefile build options here: https://github.com/libpd/libpd#androidmk-makefile-libpdxcodeproj-libpd_winsln

Looks like you are not using the EXTRA option which, if set correctly, will automatically call the setup functions for the extra externals distributed with libpd in libpd_init() which includes expr. Alternatively, you can call expr_setup(); yourself after initing libpd.

**Comment by danomatika** [comment]
Also, you need to provide what system your working on (Mac OS, Windows, etc), what layer of libpd you are using (C, Python, Obj-C, etc), how you are building it (Makefile, Xcode project, etc) for us to help understand what's going on.

**Comment by 0xJ1** [comment]
pd-for-android 1.0.0
Built using README.md terminal steps on Windows 10 via gradlew
Tested debug apk via Android Studio on Samsung Galaxy Note 3 hardware with Android 5.0 Lollipop

**Comment by danomatika** [comment]
Since it looks like a pd-for-android issue, I would open an issue there: https://github.com/libpd/pd-for-android

**Comment by danomatika** [comment]
Actually, I went ahead and opened an issue already: https://github.com/libpd/pd-for-android/issues/38

**Comment by joebowbeer** [comment]
@0xJ1 are you building pd-for-android from source or using the jcenter lib as a gradle dependency in another project?

**Comment by b2renger** [comment]
It's probably related to this one ? https://github.com/libpd/pd-for-android/issues/32

a workaround is removing the '/arm64-v8a' folder from /libs

**Comment by danomatika** [comment]
Either way, I definitely recommend adding the print concatenation layer to the build.

This is easier to read:

~~~
D/MainActivity.pdPost: expr $f1* (100.0/127.0)
D/MainActivity.pdPost: error: ... couldn't create
~~~

than this:

~~~
D/MainActivity.pdPost: expr
D/MainActivity.pdPost:

D/MainActivity.pdPost: $f1
D/MainActivity.pdPost:

D/MainActivity.pdPost: *
D/MainActivity.pdPost:

D/MainActivity.pdPost: (100.0/127.0)
D/MainActivity.pdPost: error: ... couldn't create
~~~

(IMO this should be the default behavior, but it was an early design decision.)

**Comment by 0xJ1** [comment]
@joebowbeer I've tried both building ```gradlew```, importing to Android Studio then adding a project, and building ```gradlew``` then importing the .aar into a standalone project per https://github.com/libpd/pd-for-android, both methods had issue with ```expr```.
 
Not sure if this is related at all, but on my last ```gradlew``` build I did notice this warning (should it be an error?) on PdCore:
```
[armeabi] Compile thumb  : pd <= z_hooks.c
[armeabi] Compile thumb  : choice <= choice.c
[armeabi] Compile thumb  : expr <= vexp.c
[armeabi] Compile thumb  : expr <= vexp_fun.c
[armeabi] Compile thumb  : expr <= vexp_if.c
C:/pd-for-android-aar/PdCore/jni/libpd/pure-data/extra/expr~/vexp.c: In function 'ex_eval':
C:/pd-for-android-aar/PdCore/jni/libpd/pure-data/extra/expr~/vexp.c:1114:25: warning: format '%d' expects argument of type 'int', but argument 3 has type 'long int' [-Wformat=]
                         "expr: ex_eval: unexpected type %d\n", eptr->ex_type);
                         ^
C:/pd-for-android-aar/PdCore/jni/libpd/pure-data/extra/expr~/vexp.c:1196:17: warning: format '%x' expects argument of type 'unsigned int', but argument 3 has type 'long int' [-Wformat=]
                 post_error((fts_object_t *) expr, "expr: ex_print: bad op 0x%x\n", eptr->ex_op);
                 ^
[armeabi] Compile thumb  : fiddle_tilde <= fiddle~.c
[armeabi] Compile thumb  : loop_tilde <= loop~.c
[armeabi] Compile thumb  : lrshift_tilde <= lrshift~.c
<snip>
[eabi-v7a] Compile thumb  : choice <= choice.c
[armeabi-v7a] Compile thumb  : expr <= vexp.c
[armeabi-v7a] Compile thumb  : expr <= vexp_fun.c
[armeabi-v7a] Compile thumb  : expr <= vexp_if.c
C:/pd-for-android-aar/PdCore/jni/libpd/pure-data/extra/expr~/vexp.c: In function 'ex_eval':
C:/pd-for-android-aar/PdCore/jni/libpd/pure-data/extra/expr~/vexp.c:1114:25: warning: format '%d' expects argument of type 'int', but argument 3 has type 'long int' [-Wformat=]
                         "expr: ex_eval: unexpected type %d\n", eptr->ex_type);
                         ^
C:/pd-for-android-aar/PdCore/jni/libpd/pure-data/extra/expr~/vexp.c:1196:17: warning: format '%x' expects argument of type 'unsigned int', but argument 3 has type 'long int' [-Wformat=]
                 post_error((fts_object_t *) expr, "expr: ex_print: bad op 0x%x\n", eptr->ex_op);
                 ^
[armeabi-v7a] Compile thumb  : fiddle_tilde <= fiddle~.c
[armeabi-v7a] Compile thumb  : loop_tilde <= loop~.c
[armeabi-v7a] Compile thumb  : lrshift_tilde <= lrshift~.c
[armeabi-v7a] Compile thumb  : pdnative <= z_jni_plain.c
```

**Comment by joebowbeer** [comment]
@0xJ1 I'm wondering why you're building the .aar (right?) and not using the prebuilt one published to jcenter. Ideally you shouldn't have to clone or fork or build pd-for-android.

You can find where the expr functions are specially handled in the build.gradle, by the way.

As for those warning messages, they are issues in libpd regarding the string formatting of error messages. We'll pick up fixes for them when we rev to a new version, but I've been told they are harmless for now.

**Comment by joebowbeer** [comment]
@0xJ1 By the way, if you crack open the .aar you should see the handlers for expr, expr~ and fexpr~ in the /jni/<arch> folders.

**Comment by 0xJ1** [comment]
I was able to resolve the ```expr - error: couldn't create``` issue by using the [PdTest Android Studio project] (https://github.com/libpd/pd-for-android/tree/master/PdTest) as a base build template for the project.

Looks like this issue has spawned some other features/issues, I leave it up to the group wether this should be closed out.  

From my perspective my original issue has been resolved. Thanks all for your efforts.
